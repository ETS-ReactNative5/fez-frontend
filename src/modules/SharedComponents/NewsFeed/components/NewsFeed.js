import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

import {StandardCard} from 'modules/SharedComponents/Toolbox/StandardCard';
import {locale} from 'locale';
const moment = require('moment');
const dompurify = require('dompurify');
import ReactHtmlParser from 'react-html-parser';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import {withStyles} from '@material-ui/core/styles';

const styles = theme => ({
    newsItem: {
        color: theme.palette.primary.main,
        '& .day': {
            fontSize: '1.7rem',
            fontWeight: theme.typography.fontWeightMedium
        },
        '& .month': {
            textTransform: 'uppercase',
            ontWeight: theme.typography.fontWeightMedium,
            fontSize: '1.1rem'
        },
        '& .year': {
            fontSize: '0.9rem',
            ontWeight: theme.typography.fontWeightMedium
        }
    }
});

export class NewsFeed extends PureComponent {
    static propTypes = {
        newsFeedList: PropTypes.array,
        loadingNewsFeedList: PropTypes.bool,
        showNewsCount: PropTypes.number,
        actions: PropTypes.object,
        classes: PropTypes.object
    };

    static defaultProps = {
        newsFeedList: [],
        loadingNewsFeedList: true,
        showNewsCount: 3
    };

    componentDidMount() {
        if (this.props.actions && this.props.actions.loadNewsFeed) {
            this.props.actions.loadNewsFeed();
        }
    }

    render() {
        const {classes} = this.props;
        const txt = locale.components.newsFeed;

        if (this.props.loadingNewsFeedList || this.props.newsFeedList.length === 0) {
            return null;
        }

        const allowedHtmlConfig = { ALLOWED_TAGS: ['p', 'strong', 'i', 'u', 's', 'strike', 'sup', 'sub', 'em', 'br', 'b', 'sup', 'sub'], ALLOWED_ATTR: [] };
        const subNewsFeed = this.props.newsFeedList
            .slice(0,
                this.props.newsFeedList.length > this.props.showNewsCount
                    ? this.props.showNewsCount
                    : this.props.newsFeedList.length
            );

        return (
            <StandardCard title={txt.title} darkHeader>
                {
                    !this.props.loadingNewsFeedList && subNewsFeed.map((newsItem, index) => (
                        <div style={{padding: '8px 0px'}} key={index}>
                            <Grid key={`newsItem-${index}`} container spacing={16} className={classes.newsItem}>
                                <Grid item xs={'auto'}>
                                    <Grid container direction="column" alignItems="center" justify="center" alignContent="center" spacing={0}>
                                        <Grid item className="day">{moment(newsItem.nws_updated_date).format('D')}</Grid>
                                        <Grid item className="month">{moment(newsItem.nws_updated_date).format('MMM')}</Grid>
                                        <Grid item className="year">{moment(newsItem.nws_updated_date).format('YYYY')}</Grid>
                                    </Grid>
                                </Grid>
                                <Grid item xs>
                                    <Typography>
                                        <b>{newsItem.nws_title}</b> {ReactHtmlParser(dompurify.sanitize(newsItem.nws_message, allowedHtmlConfig))}
                                    </Typography>
                                </Grid>
                            </Grid>
                        </div>
                    ))
                }
            </StandardCard>
        );
    }
}

export default withStyles(styles, {withTheme: true})(NewsFeed);
