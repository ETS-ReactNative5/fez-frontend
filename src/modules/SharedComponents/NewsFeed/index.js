export { default as NewsFeed } from './containers/NewsFeed';
