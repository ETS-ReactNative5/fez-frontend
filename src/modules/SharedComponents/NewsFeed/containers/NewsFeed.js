import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import NewsFeed from '../components/NewsFeed';
import * as actions from 'actions';

const mapStateToProps = (state) => {
    return {
        ...state.get('newsFeedReducer')
    };
};

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(actions, dispatch)
    };
}

const NewsFeedContainer = connect(mapStateToProps, mapDispatchToProps)(NewsFeed);
export default NewsFeedContainer;
