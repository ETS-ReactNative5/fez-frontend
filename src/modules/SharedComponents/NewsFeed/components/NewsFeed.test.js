import {NewsFeed} from './NewsFeed';

function setup(testProps, isShallow = true){
    const props = {
        classes: {},
        actions: {
            loadNewsFeed: jest.fn()
        },
        loadingNewsFeedList: false,
        newsFeedList: [
            {
                'nws_usr_id': 1,
                'nws_created_date': '2006-12-15 02:59:07',
                'nws_updated_date': '2018-12-15 02:59:07',
                'nws_title': 'Outage',
                'nws_status': 'active',
                'nws_message': 'There will be a 30 minute outage from 12:00-12:30pm on Wednesday 22/8/17 to upgrade the eSpace database. Apologies for this inconvenience.',
                'nws_admin_only': 0
            },
            {
                'nws_usr_id': 1,
                'nws_created_date': '2006-12-15 02:59:07',
                'nws_updated_date': '2017-12-15 02:59:07',
                'nws_title': 'Statistics',
                'nws_status': 'active',
                'nws_message': 'The file download and page view statistics for publications in eSpace currently may not be accurate and are being reviewed by our development team. Until this issue is resolved eSpace statistics should not be considered reliable.',
                'nws_admin_only': 0
            },
            {
                'nws_usr_id': 1,
                'nws_created_date': '2006-12-15 02:59:07',
                'nws_updated_date': '2016-12-15 02:59:07',
                'nws_title': 'PREFS - Account Preferences',
                'nws_status': 'active',
                'nws_message': 'When you set \'Remember search parameters\' to \'Yes\' via the PREFS tab, date filters applied to searches will now be remembered between searches',
                'nws_admin_only': 0
            },
        ],
        ...testProps
    };
    return getElement(NewsFeed, props, isShallow);
}

describe('Component NewsFeed', () => {
    it('should not render anything while loading', () => {
        const wrapper = setup({loadingNewsFeedList: true});
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not render anything if there are no news', () => {
        const wrapper = setup({newsFeedList: []});
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render news feed', () => {
        const wrapper = setup({});
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it.skip('should call action to load news feed', () => {
        const loadFn = jest.fn();
        const wrapper = setup({actions: { loadNewsFeed: loadFn}});
        wrapper.instance().componentDidMount();
        expect(loadFn).toHaveBeenCalled();
    });
});
