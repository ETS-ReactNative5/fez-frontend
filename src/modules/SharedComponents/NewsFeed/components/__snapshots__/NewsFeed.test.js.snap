// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component NewsFeed should not render anything if there are no news 1`] = `""`;

exports[`Component NewsFeed should not render anything while loading 1`] = `""`;

exports[`Component NewsFeed should render news feed 1`] = `
<StandardCard
  primaryHeader={true}
  title="Latest news"
>
  <div
    key="0"
    style={
      Object {
        "padding": "8px 0px",
      }
    }
  >
    <WithStyles(Grid)
      container={true}
      key="newsItem-0"
      spacing={16}
    >
      <WithStyles(Grid)
        item={true}
        xs="auto"
      >
        <WithStyles(Grid)
          alignContent="center"
          alignItems="center"
          container={true}
          direction="column"
          justify="center"
          spacing={0}
        >
          <WithStyles(Grid)
            className="day"
            item={true}
          >
            15
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="month"
            item={true}
          >
            Dec
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="year"
            item={true}
          >
            2018
          </WithStyles(Grid)>
        </WithStyles(Grid)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={true}
      >
        <WithStyles(Typography)
          variant="body2"
        >
          <b>
            Outage
          </b>
           
          There will be a 30 minute outage from 12:00-12:30pm on Wednesday 22/8/17 to upgrade the eSpace database. Apologies for this inconvenience.
        </WithStyles(Typography)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </div>
  <div
    key="1"
    style={
      Object {
        "padding": "8px 0px",
      }
    }
  >
    <WithStyles(Grid)
      container={true}
      key="newsItem-1"
      spacing={16}
    >
      <WithStyles(Grid)
        item={true}
        xs="auto"
      >
        <WithStyles(Grid)
          alignContent="center"
          alignItems="center"
          container={true}
          direction="column"
          justify="center"
          spacing={0}
        >
          <WithStyles(Grid)
            className="day"
            item={true}
          >
            15
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="month"
            item={true}
          >
            Dec
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="year"
            item={true}
          >
            2017
          </WithStyles(Grid)>
        </WithStyles(Grid)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={true}
      >
        <WithStyles(Typography)
          variant="body2"
        >
          <b>
            Statistics
          </b>
           
          The file download and page view statistics for publications in eSpace currently may not be accurate and are being reviewed by our development team. Until this issue is resolved eSpace statistics should not be considered reliable.
        </WithStyles(Typography)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </div>
  <div
    key="2"
    style={
      Object {
        "padding": "8px 0px",
      }
    }
  >
    <WithStyles(Grid)
      container={true}
      key="newsItem-2"
      spacing={16}
    >
      <WithStyles(Grid)
        item={true}
        xs="auto"
      >
        <WithStyles(Grid)
          alignContent="center"
          alignItems="center"
          container={true}
          direction="column"
          justify="center"
          spacing={0}
        >
          <WithStyles(Grid)
            className="day"
            item={true}
          >
            15
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="month"
            item={true}
          >
            Dec
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="year"
            item={true}
          >
            2016
          </WithStyles(Grid)>
        </WithStyles(Grid)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={true}
      >
        <WithStyles(Typography)
          variant="body2"
        >
          <b>
            PREFS - Account Preferences
          </b>
           
          When you set 'Remember search parameters' to 'Yes' via the PREFS tab, date filters applied to searches will now be remembered between searches
        </WithStyles(Typography)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </div>
</StandardCard>
`;

exports[`Component NewsFeed should show new count if above set limit 1`] = `
<StandardCard
  darkHeader={true}
  title="Latest news"
>
  <div
    key="0"
    style={
      Object {
        "padding": "8px 0px",
      }
    }
  >
    <WithStyles(Grid)
      container={true}
      key="newsItem-0"
      spacing={16}
    >
      <WithStyles(Grid)
        item={true}
        xs="auto"
      >
        <WithStyles(Grid)
          alignContent="center"
          alignItems="center"
          container={true}
          direction="column"
          justify="center"
          spacing={0}
        >
          <WithStyles(Grid)
            className="day"
            item={true}
          >
            15
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="month"
            item={true}
          >
            Dec
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="year"
            item={true}
          >
            2018
          </WithStyles(Grid)>
        </WithStyles(Grid)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={true}
      >
        <WithStyles(Typography)
          variant="body2"
        >
          <b>
            Outage
          </b>
           
          There will be a 30 minute outage from 12:00-12:30pm on Wednesday 22/8/17 to upgrade the eSpace database. Apologies for this inconvenience.
        </WithStyles(Typography)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </div>
</StandardCard>
`;
